lib/vrl/compiler/src/expression/block.rs

```diff
@@ -66,6 +66,24 @@ impl Expression for Block {
         type_def.with_fallibility(fallible)
     }
 
+    fn as_value(&self) -> Option<Value> {
+        let mut values = self
+            .inner
+            .iter()
+            .map(Expression::as_value)
+            .collect::<Vec<Option<_>>>();
+
+        // Even though we only care about the last expression in the block, we
+        // still want to make sure *all* expressions in the block are constant
+        // values, to avoid other parts of the compiler assuming the block can
+        // be optimized away, causing expressions with potential side-effects
+        // from being ignored at runtime.
```

Comment:
I realized that this is actually the case in other implementations as well. Most obvious is the assignment expression:

```
.foo = true
```

This returns a constant value, but the expression has side effects, so the compiler should never assume it can optimize this away.

We can do three things here:

1. Make sure any expression with side effects never returns `Some` for `as_value`.
2. Document that `as_value` should never be used to optimize away the original expression.
3. Have `as_value` return e.g. `Option<AsValue>` with `enum AsValue { WithSideEffects(Value), Pure(Value) }` or something similar.

Any thoughts or preferences?

Reply:
To be clear on this statement:

> compiler should never assume it can optimize this away

We're not doing anything like that right now. I'm mostly musing potential pitfalls _if_ we decide to add some kine of compiler optimization pass.